Add timeout injection to faulty agent for testing

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37485 Add timeout injection to faulty agent for testing**

Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

rohan-varma committed:
Add timeout injection to faulty agent for testing

Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -14,14 +14,16 @@ FaultyProcessGroupAgent::FaultyProcessGroupAgent(
       failNumSends_(failNumSends),
-      messageTypesToFail_(parseMessagesToFailInput(messagesToFail)) {}
+      messageTypesToFail_(parseMessagesToFailInput(messagesToFail)),
+      messagesToDelay_(parseMessagesToDelay(messagesToDelay)) {}
```
rename this to `messageTypesToDelay`? (just for consistency with `messageTypesToFail`)

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -31,20 +33,23 @@ std::vector<MessageType> FaultyProcessGroupAgent::parseMessagesToFailInput(
+    parseMessagesToDelay(
+        const std::unordered_map<std::string, float>& messagesToDelay) const {
+  std::unordered_map<MessageType, float, std::hash<int>> delayMessages;
+  for (const auto& messagePair : messagesToDelay) {
```
We should assert that all delays are non-negative

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -76,13 +81,45 @@ std::shared_ptr<FutureMessage> FaultyProcessGroupAgent::send(
+  if (msgDelay != 0) {
+    // Sleep for the specified delay for the message.
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(static_cast<int>(msgDelay * 1000)));
```
We should make this 1000 some constant like `kConvertToMilliseconds` (there's probably a better name to give this)

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -14,14 +14,16 @@ FaultyProcessGroupAgent::FaultyProcessGroupAgent(
     int numSendRecvThreads,
     std::chrono::milliseconds rpcTimeout,
     const std::vector<std::string>& messagesToFail,
+    std::unordered_map<std::string, float> messagesToDelay,
```
make this const reference like messagesToFail?

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -76,13 +81,45 @@ std::shared_ptr<FutureMessage> FaultyProcessGroupAgent::send(
 }
 
+float FaultyProcessGroupAgent::getDelayForMessage(MessageType type) const {
+  return messagesToDelay_.find(type) == messagesToDelay_.end()
```
This would end up calling `find` twice if the messages exists in the map right? It might be more efficient to use the iterator:
```
auto iter = messagesToDelay_.find(type)
return iter == messagesToDelay_.end() ? 0 : iter->second;
```

osalpekar commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -76,13 +81,45 @@ std::shared_ptr<FutureMessage> FaultyProcessGroupAgent::send(
+}
+
+// Lazily constructed map that returns string to message type mapping
+const std::unordered_map<std::string, MessageType> FaultyProcessGroupAgent::
```
We only use this function to retrieve the MessageTypes, so can we refactor this to just retrieve the MessageType instead of a map from strings to types?

osalpekar reviewed (commented):
Great fixes - glad to see the faulty agent is more broadly useful!

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -14,37 +18,44 @@ FaultyProcessGroupAgent::FaultyProcessGroupAgent(
   // types. We will then check this list of types in the send function to
   // determine whether we should fail or not.
-  std::vector<MessageType> messageTypesToFail;
+  std::vector<MessageType> messageTypesToFail(messagesToFail.size());
```
Should this be 

```c++
std::vector<MessageType> messageTypesToFail;
messageTypesToFail.reserve(messagesToFail.size());
```

Otherwise, the `push_back` below will start from `messagesToFail.size() + 1`'s item?

rohan-varma replied:
Good point, will update

mrshenli commented on torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp:
```diff
@@ -76,13 +87,44 @@ std::shared_ptr<FutureMessage> FaultyProcessGroupAgent::send(
+      {"PYTHON_CALL", MessageType::PYTHON_CALL},
+      {"SCRIPT_CALL", MessageType::SCRIPT_CALL},
+  };
+  return msgMap.find(messageString)->second;
```
do we care about absent `messageString` in the map?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2393,3 +2393,75 @@ def test_custom_faulty_messages(self):
+        with self.assertRaisesRegex(RuntimeError, expected_error):
+            fut.wait()
+
+        # Ensure timeout if we set a new default and don't override
```
should we add one more test that 1)`DEFAULT_RPC_TIMEOUT_SEC` is reasonably larger than the specified 1.5s delay 2) `fut.wait()` would succeed with default timeout?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2393,3 +2393,75 @@ def test_custom_faulty_messages(self):
+        rpc._set_rpc_timeout(rpc.constants.DEFAULT_RPC_TIMEOUT_SEC)
+
+    @dist_init(faulty_messages=[], messages_to_delay={"SCRIPT_CALL": 1.5})
+    def test_rpc_script_timeout(self):
```
Looks like tests didn't cover `PYTHON_CALL` delay?

rohan-varma replied:
python_call which corresponds to python UDF is already tested in `test_rpc_timeouts` (in the regular RPC tests, not in the faulty agent) - we test those currently by having a UDF that just `sleep`s on the remote end. We can just keep those tests, unless there's a reason to move them here as well.

mrshenli reviewed (commented):
LGTM! Just left some minor comments.

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Please wait for tests before landing.

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add timeout injection to faulty agent for testing"


Adds arbitrary timeout injection to faulty RPC agent. This is to better test scenarios that need information about how long-running RPCs, such as properly testing RPC timeouts and the profiler in all scenarios.

This is done by overriding ProcessGroupAgent's `enqueueSend()` function to inject the timeout. Determining which messages to timeout is done similar to the existing `faulty_messages` by having the user specify a mapping of message to timeout.

Added unit tests that verify RPC timeouts work with builtin + TorchScript functions, which was not tested before.

Differential Revision: [D21296537](https://our.internmc.facebook.com/intern/diff/D21296537/)

[ghstack-poisoned]
